### ISSUE TITLE
select2 add new row 

### ISSUE THREAD

**Issue by performiture** [Mottie/tablesorter#1237]
I am trying to add new row to the table with select2 filter. If you click "Add Row" the tbody will become blank until filter reset button is hit. Please see  https://jsfiddle.net/igorsf/f69jte28/

**Comment by Mottie** [comment]
Hi @performiture!

It looks like it's an issue with the encoding of filters for storage. I'll have this problem fixed shortly.

**Comment by performiture** [comment]
Hi @Mottie,

Thanks for fixing this.. I think I found another issue related to this. When I set select2 filter on the last column all rows will be filtered. Here is demo https://jsfiddle.net/igorsf/fynwfryk/2/  If you try to select a city column 5 you will not get results but if you try column 1 it works fine.

**Comment by Mottie** [comment]
Actually it wasn't related; but a fix is now available for the select2 filter formatter in the master branch.

**Comment by Mottie** [comment]
[Updated demo](https://jsfiddle.net/Mottie/fynwfryk/3/)

**Comment by performiture** [comment]
thanks for such a quick turnaround!

**Comment by Mottie** [comment]
Well, I try :frowning:... someday I'll work on an update for the select2 filter formatter  so it supports version 4+.

**Comment by performiture** [comment]
Could you take a look at $('table').trigger("updateAll") event with of select2 filter formatter. In this example I am trying to change the filter_function on click but get an error `widget-filter-formatter-select2.js:57` Uncaught TypeError: v.replace is not a function`  Here is example  https://jsfiddle.net/igorsf/pze42rhw/1/  If you click on "Change Filter" you will see this error.

**Comment by Mottie** [comment]
The issue there was that the select2 file was still targeting a DOM node that was removed. So now it adds a class name to the cell and targets that instead. There was also an issue, which was fixed, that popped up in the `widget-filter.js` file which only occurred after an "updateAll".

Anyway, I wouldn't recommend using "updateAll" because it is meant to update the table *after* content in the `thead` and `tbody` are modified - like adding or removing a column; and it still causes issues with some widgets.

The best method to update the filter widget would be to use the [`refreshWidgets` method](https://mottie.github.io/tablesorter/docs/#refreshwidgets). 

```js
$('table').trigger("refreshWidgets", [true]);
```

Here is an [updated demo](https://jsfiddle.net/Mottie/pze42rhw/3/) which is using an updated `widget-filter.js`. I usually only update the `jquery.tablesorter.widgets.js` file with each release.

**Comment by performiture** [comment]
Thank you again! Both "updateAll" and "refreshWidgets" work now as described.

It would be really great if select2 selected option was retained after refresh just like other filters. Here is a demo https://jsfiddle.net/igorsf/awueejkh/8/  If you select to filter by "10-100" and city "Los Angeles" and hit "Run" to refresh, the city filter selected option is empty.

**Comment by Mottie** [comment]
That demo is missing the `widget-storage.js` file, and it wasn't using the most up-to-date versions of the `widget-filter.js` and `widget-filter-formatter-select2.js` (the `jquery.tablesorter.widgets.js` file has not yet been updated) - here is an updated [demo](https://jsfiddle.net/Mottie/awueejkh/10/).

Also, I had to make a few more tweaks to the select2 script, so please check the master branch.

**Comment by performiture** [comment]
Mottie, I've been trying to make this work with with async ajax call to populate table rows. Here is my [demo](https://jsfiddle.net/igorsf/fy9y6x7p/2/)  When refreshing, the select2 filter (city) will go blank. Interestingly if I add `$('table').trigger("refreshWidgets", [true]);` it works in this demo but in my project I loose selected values for other filters on refresh.

**Comment by Mottie** [comment]
Hmm, I got the options to start working, but the selection isn't being saved. Give me a bit more time to work on this problem.

**Comment by Mottie** [comment]
Sorry, I haven't had a chance to work on this issue... too many distractions. I'll try to get to it soon.

**Comment by performiture** [comment]
Of course. Thank you for your fantastic work

**Comment by performiture** [comment]
Bump. Would be nice to make progress on this.

**Comment by Mottie** [comment]
Hiya! Sorry for the loooong delay...

Today I was working on finishing up the documentation on the new Mark widget. Once I get that done, this was the next thing on my to do list. I plan to have a new release available by this weekend.

**Comment by Mottie** [comment]
Ok, here is what I have figured out so far.

* When `select2` does not find a match, it clears out the value. So that gets lost before the rows are added.
* To get around that I had to save the filters at initialization; then reapply the filters after the rows are added (ajax completes).

    ```js
    $('.tablesorter')
      .on('filterInit', function() {
        savedFilters = $(this).data('lastSearch');
      })
      .tablesorter({ /* ... */ });

    // after rows are added
    $('table')
      .data('lastSearch', savedFilters)
      .trigger("update");
    ```

* When an "update" is triggered, it wasn't updating the filter formatter functions. It does now. We could have fixed the issue by triggering a "filterFomatterUpdate" on the table, but that's now done internally.
* So now the last problem.... the stickyHeader widget is interfering with the update. If you remove the stickyHeader widget, it all works perfectly.

Current demo: https://jsfiddle.net/Mottie/fy9y6x7p/4/

To Do: get stickyHeaders to cooperate.

**Comment by Mottie** [comment]
Well, it's a circular problem... I have now traced the last issue to the filter widget when it binds the sticky header clone filter inputs. It's all very frustrating and time consuming. I decided not to hold off on the release I planned for this weekend. I'll continue to isolate the problem as I find time/motivation.

**Comment by performiture** [comment]
Mottie, your demo (without sticky headers) works great if ajax takes longer to complete then tablesorter filterInit. When I tried it with real ajax call, it completed before savedFilters got initialized causing lastSearch to be overwritten with empty array.  To reproduce this you can just lower the timeout to 30ms or so from 3 sec. Is there someway I can ensure the order of events here?


 it seems I have a race condition because ajax completes before savedFilters variable is assigned  the filters at initialization.

**Comment by Mottie** [comment]
Well, there have been some issues with race-conditions. This situation is another such case. The filter widget does wait until tablesorter has completed initialization before applying the saved search query. Triggering an "update" before the filter widget completes initialization does interfere with that sequence.

Moving the code to store the saved filters to the `intialized` callback function, appears to be a more reliable method, *but* now the select2 input isn't updating as expected: https://jsfiddle.net/fy9y6x7p/5/

**Comment by performiture** [comment]
Is it possible to execute ajax call that triggers an update only after tablesorter has initialized?

**Comment by Mottie** [comment]
You may not need the `setTimeout` if you add that ajax call to the [`initialized` callback function](https://mottie.github.io/tablesorter/docs/#initialized).

**Comment by performiture** [comment]
The problem with moving ajax call to `initialized` is as you pointed out earlier select2 input not updating as expected: https://jsfiddle.net/igorsf/fy9y6x7p/7/

**Comment by Mottie** [comment]
Ok, I think I got it *mostly* fixed now... I needed to add a slight delay before updating the filter formatter functions.

So now [this demo works](https://jsfiddle.net/fy9y6x7p/9/), and it even works when the stickyHeaders widget is included; but the select2 within the sticky header isn't updated LOL.... one step at a time.

**Comment by performiture** [comment]
I noticed you are setting but not using savedFilters variable in the demo. It works but as soon as I add timeout to simulate ajax request, it breaks [demo](https://jsfiddle.net/igorsf/fy9y6x7p/10/)

**Comment by Mottie** [comment]
Oops, I should have fixed that... try [this demo](https://jsfiddle.net/fy9y6x7p/11/).

**Comment by performiture** [comment]
It now works perfectly with "select2" input. However in my app, 1 out of 4 refreshes, the columns with "filter-select" class are loosing their selected values. I tried debugging it. When `update` is triggered and getFilters function in widget-filter.js is called, it has the right filters in `$( table ).data( 'lastSearch' )` but after the function ends the `filters` variable array that it returns is missing those "filter-select" columns.

**Comment by Mottie** [comment]
If the select2 input isn't updating, try triggering a `filterFomatterUpdate` event on the table. The `setTimeout` I recently added is set to 100ms, maybe it needs more time.

**Comment by performiture** [comment]
select2 is updating just fine. Its the other filters like `<th class="filter-select" data-placeholder="Select a name">First Name</th>` which sometimes don't get set.

**Comment by Mottie** [comment]
I can't duplicate that problem. Please modify [this demo](https://jsfiddle.net/Mottie/fy9y6x7p/14/) to show me the problem so I can troubleshoot it better.
